packages/api/src/middleware/validation/checkAnimalOrBatch.js

```diff
@@ -13,81 +13,622 @@
+    customTypeId = animalOrBatchRecord.custom_type_id;
+  }
+
+  // Custom breed does not match type if defaultId OR customTypeId does not match
+  if (
+    (customBreed.default_type_id && customBreed.default_type_id !== defaultTypeId) ||
+    (customBreed.custom_type_id && customBreed.custom_type_id !== customTypeId)
+  ) {
+    throw customError('Breed does not match type');
+  }
+};
+
+const checkAnimalBreed = async (
+  animalOrBatch,
+  farm_id,
+  animalOrBatchRecord = undefined,
+  creating = true,
+) => {
+  const {
+    default_breed_id,
+    custom_breed_id,
+    breed_name,
+    default_type_id,
+    custom_type_id,
+    type_name,
+  } = animalOrBatch;
+  const breedKeyOptions = ['default_breed_id', 'custom_breed_id', 'breed_name'];
+  const typeKeyOptions = ['default_type_id', 'custom_type_id', 'type_name'];
+
+  // Check if breed is present
+  if (
+    (creating && oneExists(breedKeyOptions, animalOrBatch)) ||
+    oneTruthy([default_breed_id, custom_breed_id, breed_name])
+  ) {
+    checkExactlyOneIsProvided(
+      [default_breed_id, custom_breed_id, breed_name],
+      'default_breed_id, custom_breed_id, or breed_name',
+    );
+  }
+  // Check if breed is present
+  if (!creating && oneExists(breedKeyOptions, animalOrBatch)) {
+    // Overwrite with null in db if editing
+    setFalsyValuesToNull(breedKeyOptions, animalOrBatch);
+  }
+
+  if (
+    oneExists(breedKeyOptions, animalOrBatch) &&
+    !oneTruthy([default_breed_id, custom_breed_id, breed_name])
+  ) {
+    // do nothing if nulling breed
+  } else {
+    // Check if default breed or default type is present
+    if (
+      (oneExists(breedKeyOptions, animalOrBatch) && default_breed_id) ||
+      (oneExists(typeKeyOptions, animalOrBatch) && default_type_id)
+    ) {
+      await checkDefaultBreedMatchesType(animalOrBatchRecord, default_breed_id, default_type_id);
+    }
+    // Check if custom breed or custom type is present
+    if (
+      (oneExists(breedKeyOptions, animalOrBatch) && custom_breed_id && !type_name) ||
+      (oneExists(typeKeyOptions, animalOrBatch) &&
+        (default_type_id || custom_type_id) &&
+        !breed_name)
+    ) {
+      let customBreed;
+      // Find customBreed if exists
+      if (oneExists(breedKeyOptions, animalOrBatch) && custom_breed_id) {
+        checkIdIsNumber(custom_breed_id);
+        customBreed = await CustomAnimalBreedModel.query()
+          .whereNotDeleted()
+          .findById(custom_breed_id);
+        if (!customBreed) {
+          throw customError('Custom breed does not exist');
+        }
+      } else if (animalOrBatchRecord?.custom_breed_id) {
+        checkIdIsNumber(animalOrBatchRecord?.custom_breed_id);
+        customBreed = await CustomAnimalBreedModel.query()
+          .whereNotDeleted()
+          .findById(animalOrBatchRecord.custom_breed_id);
+        if (!customBreed) {
+          // This should not be possible
+          throw customError('Custom breed does not exist');
+        }
```

Comment:
I kept checkIdIsNumber to be consistent with other logic --- but it could be handled by objection by throwing 500 or other objection error.

I could probably remove the second check since if it is from an existing `animalOrBatchRecord` -- it would be really rare for it not to exist (concurrency)

Reply:
`animalOrBatchRecord?.custom_breed_id` is the id already stored in the DB, isn't it? I thought it wasn't necessary to validate saved data. The concurrency is a good point, let's keep it!

Reply:
Yes it is the id stored in db -- should be mostly safe to remove if we prefer that. Unlikely but possible the animal is deleted in between record retrieval and this check, likelihood increases with number of animals being edited at once. 

I can go either way -- really I would like to see `trx` used across middleware and controller to solve the issue instead of what I have here.

Reply:
Should I remove all `checkIdIsNumbers`?

Reply:
I thought this was the only place validating the data in the DB.

> possible the animal is deleted in between record retrieval and this check
This raises a question... where are you checking if the animal is deleted?
Even if you run `checkIdIsNumber`, it wouldn't tell you if the animal was deleted, right...? Also, if that's a concern, should we add a check for the animal's status right before editing?

Reply:
> This raises a question... where are you checking if the animal is deleted?

`getRecordIfExists()` checks for the animal record and adds it to `invalidIds[]` if not present using `whereNotDeleted()`. 

> Even if you run checkIdIsNumber, it wouldn't tell you if the animal was deleted, right...? 

`checkIdIsNumber()` I do not think does much. It only early returns a preferred response code and message instead of the Objection 500 server error that would eventually result from update failure.

CheckIdIsNumber and InvaldIds check are both carried over from other endpoints. I was hesitant to change the existing order of checks or to add extra code or spend more time. Personally I would consider allowing Objection to handle the error from `checkIdIsNumber()`, accept the 500 error, and delete all instances and tests. Also I would consider moving `checkInvalidIds()` into the loop --- but I think invalidIds was maybe a step towards partial success. Also custom type and breeds I think could be more in the loop too. If you think any of these changes are valuable I will happily add them. 

> Also, if that's a concern, should we add a check for the animal's status right before editing?

Yeah we could do that too... I would prefer figuring out transactions to just keep every single check query and patch action atomic. But it is harder.. it would be nice if `whereNotDeleted()` worked with insert/upserts. Should I reuse getRecordIfExists right before upsert? I do not think concurrency is a big problem but it could help.